src/client/dfuse/il/int_read.c

```diff
@@ -72,29 +66,43 @@ read_bulk(char *buff, size_t len, off_t position, struct fd_entry *entry, int *e
 ssize_t
 ioil_do_pread(char *buff, size_t len, off_t position, struct fd_entry *entry, int *errcode)
 {
-	return read_bulk(buff, len, position, entry, errcode);
+	d_iov_t     iov = {};
+	d_sg_list_t sgl = {};
+
+	sgl.sg_nr = 1;
+	d_iov_set(&iov, (void *)buff, len);
+	sgl.sg_iovs = &iov;
+
+	return read_bulksgl(&sgl, len, position, entry, errcode);
 }
 
 ssize_t
 ioil_do_preadv(const struct iovec *iov, int count, off_t position, struct fd_entry *entry,
 	       int *errcode)
 {
-	ssize_t bytes_read;
+	d_iov_t    *diov;
+	d_sg_list_t sgl        = {};
 	ssize_t total_read = 0;
 	int     i;
+	int         rc;
+
+	D_ALLOC_ARRAY(diov, count);
+	if (diov == NULL) {
+		*errcode = ENOMEM;
+		return -1;
+	}
 
 	for (i = 0; i < count; i++) {
-		bytes_read = read_bulk(iov[i].iov_base, iov[i].iov_len, position, entry, errcode);
+		d_iov_set(&diov[i], iov[i].iov_base, iov[i].iov_len);
+		total_read += iov[i].iov_len;
+	}
 
-		if (bytes_read == -1)
-			return (ssize_t)-1;
+	sgl.sg_nr   = count;
+	sgl.sg_iovs = diov;
 
-		if (bytes_read == 0)
-			return total_read;
+	rc = read_bulksgl(&sgl, total_read, position, entry, errcode);
```

Comment:
hmm im surprised that this works in case we are doing async read/write with ev.
the sgl and the iovec cannot be on the stack or freed after this function and should be on the heap and freed in the callback i believe.

Reply:
does the ioil ever do asynchronous I/O at present?  Regardless, I can allocation the sgl as well.  But it's been like that forever

Reply:
I thought that for arrays at least it copies the sgl as necessary when you submit it?

Reply:
Actually, even if it uses an event, it loops waiting for the event so I don't see an issue with using the stack in that case.

Reply:
no it does not copy the sgl in most cases. and yes we use the async API in most cases to utilize multiple contexts from different threads (this was newly added).
but yes we wait for that right after the call. i actually confused this IO from the one with dfuse that does progress in a different thread, which i was thinking the problem was, but here i should be OK. sorry for the noise